pandas/core/groupby.py

```diff
@@ -1487,19 +1489,32 @@ def aggregate(self, values, how, axis=0):
                 raise NotImplementedError
             out_shape = (self.ngroups,) + values.shape[1:]
 
-        if is_numeric_dtype(values.dtype):
-            values = com.ensure_float(values)
-            is_numeric = True
-            out_dtype = 'f%d' % values.dtype.itemsize
-        else:
-            is_numeric = issubclass(values.dtype.type, (np.datetime64,
-                                                        np.timedelta64))
+        if _is_datetime_or_timedelta_dtype(values.dtype):
+            values = values.view('int64')
+        elif is_bool_dtype(values.dtype):
+            values = _algos.ensure_float64(values)
+
+        is_numeric = is_numeric_dtype(values.dtype)
+        upsample = self.ngroups > values.size
+
+        if is_numeric and upsample:
```

Comment:
So messing with this tonight I ran into the same problems I hit the first time, and I think I'm missing something. The test that exposes this is `pandas/tseries/tests/test_resample.py:TestResamplePeriodIndex.test_fill_method_and_how_upsample`.

Without the upsample check that forces the float conversion the result from the cython function is an ndarray of integers:

```
In [1]: result
Out[1]:
array([[                   0],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   1],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   2],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   3],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   4],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   5],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   6],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   7],
       [-9223372036854775808],
       [-9223372036854775808],
       [                   8]])
```

Obviously the fillna that runs in resample.py won't work on this. Is there some `np.ma.masked` magic that I need to be using here to get the iNaT fill-value that you mention above?